Revise perf scalability doc (#367)
openservicemesh/osm-docs: content/docs/guides/ha_scale/perf-scale.md

## Testing environment

> This test is performed on a Kubernetes cluster provisioned and managed by [Microsoft Azure Kubernetes Service](https://azure.microsoft.com/en-us/services/kubernetes-service/). The cluster has 100 Standard_DS2_v2 nodes (each node has 2 vCPUs and 7GB memory) and 3 Standard_D32s_v3 nodes (32 vCPUs and 128GB memory) reserved for high intensity tests. It is ensured that there is only one OSM controller pod running with default resource quota during the test. OSM is configured in non-permissive traffic mode.

Reviewer: what does 3 nodes reserved for high intensity tests mean?

Author: This special high intensity nodepool was created and only has 3 nodes due to my resource quota of the cluster.

I've just upscaled my cluster. We will re-run the tests with larger nodes and cluster.

</p>

## Performance

In this test, we focus on the latency overhead and resource consumption added by the OSM sidecar. We currently test the per pod RPS at 100, 500, 1000, 2000 and 4000 respectively.

### Latency

#### Test 1: Relation with RPS

In our test, regardless of the RPS level, OSM adds around 0.5ms and 5ms overhead to the 50th percentile and the 99th percentile latency respectively. This is caused by the request processing in both the client and server sidecars, for tasks like mTLS authentication, metrics collection etc. Our test setup was with a single client and server. Performance may vary with more complex applications, or with different traffic policies set on the mesh.

#### Test 2: Relation with number of connections

A similar setup to RPS scenario is used to test the relationship between latency and number of connections to the sidecar proxy. In this test, the per pod RPS is set at 2500. As the number of connections increased, the request latency showed a proportional increase while the sidecar memory usage to maintain those open connections only slightly increased. The sidecar consumes around 5MB memory for every 100 connections maintained.

<p align="center">

>   <img src="/docs/images/perf/latency-conns.png" width="650"/>

Reviewer: this graph doesn't tell a great story, can we try this with a bigger node pool?

On the data plane, there is no obvious difference in terms of CPU usage. However, the memory usage is much higher in permissive mode. This is expected as more outbound cluster information is stored in the sidecar. The chart below shows the sidecar memory usage when 10 and 20 services are added in permissive mesh and non-permissive mesh respectively. Compared to the non-permissive mode, the memory usage is more than doubled.

<p align="center">
  <img src="/docs/images/perf/perm-memory.png" width="650"/>
</p>

The request latency from load generator to workloads is lower in permissive mode. This reduction is due to the RBAC policy filter path being a single wildcard filter as described above. This filter path is shorter than non-permissive mode and there is less rule matching operations for incoming requests. In our test, we observe the P99 latency drops to around 80% in permissive mode. The more SMI traffic policies added to services in non-permissive mode, the larger request time overhead we expect. Below is the latency comparison between permissive and non-permissive mode, tested with 250 per pod RPS:

<p align="center">

>   <img src="/docs/images/perf/latency-perm-vs-non-perm.png" width="650"/>

Reviewer: similarly these are untenable results. i'm virtually positive that there is a bottleneck unrelated to OSM that is yielding these results.

Why can't we run this on a large cluster?

By looking at the collected metrics, in our testing scenarios, resource consumption of data plane sidecars does not change much along with the mesh size. At stable time, the average CPU usage is close to 0 and average memory usage is around 50MB. CPU usage is related to request processing and memory usage is related the size of Envoy configuration stored locally. These two factors do not change with the mesh scale.

### Other Scenarios

> #### Permissive mode

Reviewer: is the image `latency-ingress.png` used for anything?

Author: Yes. The last section tests performance in a `IngressBackend` ingress mode.

> When permissive mode is enabled in OSM, services are allowed to communicate with any other services in the mesh without additional traffic policies. This is implemented by adding an allow-all wildcard RBAC policy in the inbound listener and all known services as outbound clusters in the Envoy sidecar. The same load test described above is performed on a service mesh with permissive mode enabled.

Reviewer: Add that permissive mode still enables mTLS, and requires the services be within the mesh

> On the control plane, the CPU and memory usage both increase compared to non-permissive mode.

Reviewer 2: which are you saying:
- CPU and memory usage is higher when running permissive mode?
- CPU and memory usage rates increased at the same rate between permissive and non-permissive mode so their overall usage is the same?

Author: I don't quite understand the second point. Their overall usage cannot be the same.

I mean, for two same systems except the permissive mode config, the one with permissive mode enabled consumes more CPU and memory than the one disabled it.

> On the control plane, the CPU and memory usage both increase compared to non-permissive mode.

Reviewer 2: which are you saying:
- CPU and memory usage is higher when running permissive mode?
- CPU and memory usage rates increased at the same rate between permissive and non-permissive mode so their overall usage is the same?

On the data plane, there is no obvious difference in terms of CPU usage. However, the memory usage is much higher in permissive mode. This is expected as more outbound cluster information is stored in the sidecar. The chart below shows the sidecar memory usage when 10 and 20 services are added in permissive mesh and non-permissive mesh respectively. Compared to the non-permissive mode, the memory usage is more than doubled.

<p align="center">
  <img src="/docs/images/perf/perm-memory.png" width="650"/>
</p>

> The request latency from load generator to workloads is lower in permissive mode. This is because that the RBAC policy filter path is shorter, which is a single wildcard filter as described above. There is less rule matching process for incoming requests. In our test, we observe the P99 latency drops to around 80% in permissive mode. The more SMI traffic policies added to services in non-permissive mode, larger request time overhead we expect. Below is the latency comparison between permissive and non-permissive mode, tested with 250 per pod RPS:

Reviewer 2: ```suggestion
The request latency from load generator to workloads is lower in permissive mode. This reduction is due to the RBAC policy filter path being a single wildcard filter as described above. This filter path is shorter than non-permissive mode and there is less rule matching operations for incoming requests. In our test, we observe the P99 latency drops to around 80% in permissive mode. The more SMI traffic policies added to services in non-permissive mode, the larger request time overhead we expect. Below is the latency comparison between permissive and non-permissive mode, tested with 250 per pod RPS:
```

> The request latency from load generator to workloads is lower in permissive mode. This is because that the RBAC policy filter path is shorter, which is a single wildcard filter as described above. There is less rule matching process for incoming requests. In our test, we observe the P99 latency drops to around 80% in permissive mode. The more SMI traffic policies added to services in non-permissive mode, larger request time overhead we expect. Below is the latency comparison between permissive and non-permissive mode, tested with 250 per pod RPS:

Reviewer 2: ```suggestion
The request latency from load generator to workloads is lower in permissive mode. This reduction is due to the RBAC policy filter path being a single wildcard filter as described above. This filter path is shorter than non-permissive mode and there is less rule matching operations for incoming requests. In our test, we observe the P99 latency drops to around 80% in permissive mode. The more SMI traffic policies added to services in non-permissive mode, the larger request time overhead we expect. Below is the latency comparison between permissive and non-permissive mode, tested with 250 per pod RPS:
```

<p align="center">
  <img src="/docs/images/perf/latency-perm-vs-non-perm.png" width="650"/>
</p>

### Ingress traffic

In this test, the load generator is deployed outside of the mesh. The requests enter the mesh via [IngressBackend](https://release-v1-1.docs.openservicemesh.io/docs/guides/traffic_management/ingress/#ingressbackend-api). Therefore, the request only goes through one Envoy sidecar instead of two.

We compared three scenarios:

* Normal: load generator is inside the mesh. Request goes through 2 sidecars.
* Ingress: load generator is outside the mesh. Request goes through 1 sidecar.
* No OSM installed: Request does not go through any sidecar.

> The test result shows that, the latency of the ingress scenario drops a lot compared to that in the normal scenario. The difference is roughly equals to the latency of ingress scenario - the latency when OSM is not installed, which is the overhead added by a single Envoy sidecar. This complies with our assumption.

Reviewer 2: ```suggestion
The test result shows that the latency of the ingress scenario is significantly lower than the normal scenario. The ingress scenario is roughly equal to the results of the scenario without OSM installed. This result shows that the Envoy sidecar does add latency overhead.
```

> The test result shows that, the latency of the ingress scenario drops a lot compared to that in the normal scenario. The difference is roughly equals to the latency of ingress scenario - the latency when OSM is not installed, which is the overhead added by a single Envoy sidecar. This complies with our assumption.

Reviewer 2: ```suggestion
The test result shows that the latency of the ingress scenario is significantly lower than the normal scenario. The ingress scenario is roughly equal to the results of the scenario without OSM installed. This result shows that the Envoy sidecar does add latency overhead.
```
